src/query/functions/tag/join.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tag
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/models"
+)
+
+// TagJoinType joins the values of given tags using a given separator
+// and adds them to a given destination tag. It can combine any number of tags.
+// NB: This will override an existing tag if the given tag exists in the tag list.
+const TagJoinType = "label_join"
+
+func combineTagsWithSeparator(name []byte, separator []byte, values [][]byte) models.Tag {
+	l := len(values)
+	if l == 1 {
+		return models.Tag{Name: name, Value: values[0]}
+	}
+
+	sepLen := len(separator)
+	// initialize length to account for separators.
+	combinedLength := (l - 1) * sepLen
+	for _, v := range values {
+		combinedLength += len(v)
+	}
+
+	b := make([]byte, combinedLength)
+	idx := 0
+	for i, v := range values {
+		copy(b[idx:], v)
+		idx += len(v)
+		if i < l {
+			copy(b[idx:], separator)
+			idx += sepLen
+		}
+	}
+
+	return models.Tag{Name: name, Value: b}
+}
+
+// Gets tag values from models.
+// NB: duplicate tag names giving duplicate values is valid.
+func tagsInOrder(names [][]byte, tags models.Tags) [][]byte {
+	orderedTags := make([][]byte, 0, len(names))
+	for _, name := range names {
+		for _, tag := range tags {
+			if bytes.Equal(name, tag.Name) {
+				orderedTags = append(orderedTags, tag.Value)
+			}
+		}
+	}
+
+	return orderedTags
+}
+
+func uniqueNameCount(names []string) int {
```

Comment:
maybe move this to utils and have it return uniqueMap ? The caller can just take len(uniqueMap). This will make it more usable going forward

Reply:
Sounds good